dd_extract: Add support for compressed kernel modules

pjgeorg opened the pull request:
This PR adds support for handling compressed kernel modules to dd_extract. In particular for *.ko.gz, *.ko.xz, *.ko.bz2, *.ko.zst (these seem to be the common file extensions also supported by rpm macros). It has been opened against master to ensure the feature will be available in rhel-10 (it is of low interest for Fedora), but I plan to later also open PR against rhel-9 (and maybe rhel-8?).

I have not been able to run the full test suite locally, but instead tested dd_extract manually. I do hope the tests may be run as part of this PR.

Note that I only extended the dd_extract tests to verify it is working for xz compressed kernel modules. Shall I add cases for the other 3 file extensions as well?

Please let me know if anything seems wrong or if I forgot anything you need, first time contributing to anaconda.

VladimirSlavik commented on tests/unit_tests/dd_tests/test_dd.py:
```diff
@@ -286,7 +290,7 @@ def test_dd_extract_chmod(self):
     def test_dd_extract_modules(self):
         """dd_extract: using --modules extracts only .ko files"""
```
Please update also this docstring

VladimirSlavik commented on tests/unit_tests/dd_tests/test_dd.py:
```diff
@@ -129,6 +129,10 @@ def create_source_file(self):
     path="/lib/modules/KERNELVER/extra/net/fun.ko",
     contents="KERNEL MODULE??? YOU BETCHA"
 )
+koxzfile = TextRPMFile(
+    path="/lib/modules/KERNELVER/extra/net/fun.ko.xz",
+    contents="XZ COMPRESSED KERNEL MODULE??? YOU BETCHA"
+)
```
To achieve complete coverage, test also the 7 letter code path.

VladimirSlavik reviewed (approved):
Thank you!

I assume you tested this - does this allow actually loading compressed modules? I'm not sure how many pieces are in place to actually use them, if they now work, that's fantastic.

It would be good to perhaps also update the docs:
- A release note would be great. See `docs/release-notes/template.rst`. 
- Maybe take a look at `dracut/README-driver-updates.md` and `docs/driverdisc.rst` if there is anything to change? I don't think so, but I could be wrong.

VladimirSlavik commented:
PS: Testing farm is broken, disregard that status.

VladimirSlavik commented:
/kickstart-test --testtype smoke

VladimirSlavik commented:
/kickstart-test --testtype driverdisk

VladimirSlavik commented:
One more thought, @pjgeorg you reference a RHEL bug in the commits. Do you want us to port this change there? I can't guarantee this will happen, RHEL 8 is in late lifecycle phase. But having this upstream shifts it towards possible at least.

Note to self/team - I assume we would want to extend the driverdisk kickstart tests to include also this, check `write_kmod_rpm` in `lib/mkdud.py`

pjgeorg commented:
> I assume you tested this - does this allow actually loading compressed modules? I'm not sure how many pieces are in place to actually use them, if they now work, that's fantastic.

I have not been able to do a full test yet. Hence there might be other missing pieces. While I'm quite used to working with RPMs by now, creating .iso files is something I have never done so far. I'll look into it to be able to do a full test.
Adding support to dd_extract was the most obvious missing piece. By looking at the code I have not been able to identify any other obvious missing piece in Anaconda.
 
> It would be good to perhaps also update the docs:
> 
> * A release note would be great. See `docs/release-notes/template.rst`.
> * Maybe take a look at `dracut/README-driver-updates.md` and `docs/driverdisc.rst` if there is anything to change? I don't think so, but I could be wrong.

I'll check the docs and add appropriate changes to this PR.

> One more thought, @pjgeorg you reference a RHEL bug in the commits. Do you want us to port this change there? I can't guarantee this will happen, RHEL 8 is in late lifecycle phase. But having this upstream shifts it towards possible at least.

My main interest is in RHEL/CentOS. So, yes, I'd like to see required changes be ported once it has been merged into master. If not for RHEL 8 than maybe at least for RHEL 9. As noted in my initial message I'm happy to open necessary PRs and do the porting myself once we are there.

VladimirSlavik commented:
/boot-iso

VladimirSlavik commented:
Re RHELs, I discussed the timing with our team lead @jkonecny12 and we realized there's practically no way to get this into RHEL 8. We're too far into the 8.9 preparations to get it in, especially if it's only part of the solution. Then, 8.10 will be heavily focused on stability, because there will be no opportunity to fix anything - that's the last release and media is not rebuilt unless the sky is falling, so effectively installer does not get z-stream. Sorry. Still fine for RHEL 9 though.

VladimirSlavik commented:
I found #4191 which is definitely related, so perhaps it should work now?

pjgeorg commented:
> Re RHELs, I discussed the timing with our team lead @jkonecny12 and we realized there's practically no way to get this into RHEL 8. We're too far into the 8.9 preparations to get it in, especially if it's only part of the solution. Then, 8.10 will be heavily focused on stability, because there will be no opportunity to fix anything - that's the last release and media is not rebuilt unless the sky is falling, so effectively installer does not get z-stream. Sorry. Still fine for RHEL 9 though.

That seems reasonable. It'd be great if we could get this into RHEL 9.

> I found #4191 which is definitely related, so perhaps it should work now?

I'm not sure #4191 is related. Anyway, thanks to the boot.iso you triggered to be created I have been able to verify that it is indeed working now by using that boot.iso. I only used a simple Driver Disc with a single compressed kernel module but that should be sufficient.

pjgeorg commented:
I updated the PR:
* adding a test for the 7 letter case
* updated all docs mentioning .ko only so far
* added release notes.

 Hopefully this addresses all open issues. In case I shall change anything or anything is missing, please let me know.

VladimirSlavik commented:
/kickstart-test --testtype smoke

VladimirSlavik commented:
/kickstart-test --testtype driverdisk

VladimirSlavik commented:
/packit test

VladimirSlavik commented:
Thank you! I have to say this was a "perfect contribution" - code, tests, docs, everything.

I'll leave it for a while still so that other team members get the opportunity to comment, but IMO it's ready to merge as is.

If you have anything to say about how this went, I'm all ears.
